### PR TITLE
Loot table manipulation API

### PR DESCRIPTION
**Pull request by asiekierka**
This follows on the work laid out in #77 with some changes:

* Adapted to new directory format
* The mixins now use Accessors
* LootUtilities -> LootJsonParser with some naming clean-up
* LootEntryTypeRegistry is now a part of api.loot.v1 (with deprecation)
* LootTableLoadingCallback now receives the LootManager instance, allowing to get other loot tables than one's own.
* LootTableLoadingCallback now applies modifications at the end of all processing, which means LootManager can be read from safely.
* LootTableLoadingCallback received a setter, allowing you to override the loot table instance being iterated on completely. I am not sure if that is a good idea, but I think that's the only way we can faciliate removals at this time?

### CODE REVIEW

**Comment by asiekierka** [comment]
@Juuxel

File: fabric-testmods/java/net/fabricmc/fabric/loot/LootTableMod.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraft.world.loot.entry.LootEntry;
+
+public class LootTableMod implements ModInitializer {
+	private static final String LOOT_ENTRY_JSON = "{\"type\":\"minecraft:item\",\"name\":\"minecraft:apple\"}";
+
+	@Override
+	public void onInitialize() {
+		LootTableLoadingCallback.EVENT.register((manager, id, supplier) -> {
+			if ("minecraft:blocks/dirt".equals(id.toString())) {
+				LootEntry entryFromString = LootUtilities.readEntryFromJson(LOOT_ENTRY_JSON);
```
Comment (Juuxel) [inline fabric-testmods/java/net/fabricmc/fabric/loot/LootTableMod.java]: I think this needs to be updated for the `LootUtilities` rename.

Comment (asiekierka) [reply fabric-testmods/java/net/fabricmc/fabric/loot/LootTableMod.java]: Not yet - fabric-testmods is currently non-functional.

Comment (Juuxel) [reply fabric-testmods/java/net/fabricmc/fabric/loot/LootTableMod.java]: Ah, okay.

**Comment by asiekierka** [comment]
Any concerns about the added features?

**Comment by Juuxel** [comment]
> LootTableLoadingCallback received a setter, allowing you to override the loot table instance being iterated on completely.

I'm not 100% sure about this. It might not be needed because of data packs.

> - LootUtilities -> LootJsonParser
> - LootTableLoadingCallback now receives the LootManager instance, allowing to get other loot tables than one's own.

:+1:

---

Have you tested that the API still works?

**Comment by asiekierka** [comment]
>Have you tested that the API still works?

Not yet.

>I'm not 100% sure about this. It might not be needed because of data packs.

The use case I have in mind is transforming the loot table in such a way that certain entries are removed.

**Comment by Juuxel** [comment]
> The use case I have in mind is transforming the loot table in such a way that certain entries are removed.

Ah. I think you could alternatively add `FabricLootSupplierBuilder.removeEntry` for that, but that would require finding/building a matching entry.

File: fabric-loot-tables/src/main/java/net/fabricmc/fabric/api/loot/v1/LootJsonParser.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		} catch (Exception e) {
+			throw new RuntimeException("Exception while getting Gson instance from LootManager", e);
+		}
+	});
+
+	private LootJsonParser() {
+
+	}
+
+	private static Gson gson() {
```
Comment (Juuxel) [inline fabric-loot-tables/src/main/java/net/fabricmc/fabric/api/loot/v1/LootJsonParser.java]: This is not needed anymore, it was originally a public utility method for accessing the Gson instance.

**Comment by liach** [comment]
Can we add loot table tags for easy retrieval of multiple loot tables?

**Comment by Prospector** [comment]
This needs a slight update to fix the settings.gradle issue but otherwise should be merged if no one objects soon.

**Pull request closed without merging**